bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/util/ShellyChannelCache.java

```diff
@@ -0,0 +1,133 @@
+    private String thingName = "";
+    private boolean enabled = false;
+
+    public ShellyChannelCache(ShellyBaseHandler thingHandler) {
+        this.thingHandler = thingHandler;
+        setThingName(thingHandler.thingName);
+    }
+
+    public void setThingName(String thingName) {
+        this.thingName = thingName;
+    }
+
+    public boolean isEnabled() {
+        return enabled;
+    }
+
+    public void enable() {
+        enabled = true;
+    }
+
+    public void disable() {
+        enabled = false;
+    }
+
+    /**
+     * Update one channel. Use Channel Cache to avoid unnecessary updates (and avoid
+     * messing up the log with those updates)
+     *
+     * @param channelId Channel id
+     * @param value Value (State)
+     * @param forceUpdate true: ignore cached data, force update; false check cache of changed data
+     * @return true, if successful
+     */
+    public boolean updateChannel(String channelId, State newValue, Boolean forceUpdate) {
+        try {
+            State current = null;
+            if (channelData.containsKey(channelId)) {
+                current = channelData.get(channelId);
+            }
+            if (!enabled || forceUpdate || (current == null) || !current.equals(newValue)) {
+                if ((current != null) && current.getClass().isEnum() && (current == newValue)) {
+                    return false; // special case for OnOffType
+                }
+                // For channels that support multiple types (like brightness) a suffix is added
+                // this gets removed to get the channelId for updateState
+                thingHandler.publishState(channelId, newValue);
+                if (current == null) {
+                    channelData.put(channelId, newValue);
+                } else {
+                    channelData.replace(channelId, newValue);
+                }
+                logger.debug("{}: Channel {} updated with {} (type {}).", thingName, channelId, newValue,
+                        newValue.getClass());
+                return true;
+            }
+        } catch (IllegalArgumentException e) {
+            logger.debug("{}: Unable to update channel {} with {} (type {}): {} ({})", thingName, channelId, newValue,
+                    newValue.getClass().toString(), ShellyUtils.getMessage(e), e.getClass().toString());
+        }
+        return false;
+    }
+
+    public boolean updateChannel(String group, String channel, State value) {
+        return updateChannel(mkChannelId(group, channel), value, false);
+    }
+
+    public boolean updateChannel(String channelId, State value) {
+        return updateChannel(channelId, value, false);
+    }
+
+    /**
+     * Get a value from the Channel Cache
+     *
+     * @param group Channel Group
+     * @param channel Channel Name
+     * @return the data from that channel
+     */
+    @Nullable
+    public Object getValue(String group, String channel) {
```

Comment:
```suggestion
    public @Nullable State getValue(String group, String channel) {
```

In addition would it be possible to get rid of the nullable return value by returning `UndefType.NULL` instead?

Reply:
yep, changed

Reply:
changed

Reply:
changed